src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/DefaultRazorFormattingService.cs

```diff
@@ -54,9 +65,90 @@ internal class DefaultRazorFormattingService : RazorFormattingService
             _server = server;
             _csharpFormatter = new CSharpFormatter(documentMappingService, server, filePathNormalizer);
             _htmlFormatter = new HtmlFormatter(server, filePathNormalizer);
+            _optionsMonitor = optionsMonitor;
             _logger = loggerFactory.CreateLogger<DefaultRazorFormattingService>();
         }
 
+        public override Task<TextEdit[]> FormatOnTypeAsync(Uri uri, RazorCodeDocument codeDocument, Position position, string character, FormattingOptions options)
+        {
+            if (uri is null)
+            {
+                throw new ArgumentNullException(nameof(uri));
+            }
+
+            if (codeDocument is null)
+            {
+                throw new ArgumentNullException(nameof(codeDocument));
+            }
+
+            if (position is null)
+            {
+                throw new ArgumentNullException(nameof(position));
+            }
+
+            if (string.IsNullOrEmpty(character))
+            {
+                throw new ArgumentNullException(nameof(character));
+            }
+
+            if (options is null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (!_optionsMonitor.CurrentValue.AutoClosingTags || character != ">")
+            {
+                // We currently only support auto-closing tags our onType formatter.
+                return Task.FromResult(EmptyArray);
+            }
+
+            bool addCursorPlaceholder;
+            if (options.TryGetValue(LanguageServerConstants.ExpectsCursorPlaceholderKey, out var value) && value.IsBool)
+            {
+                addCursorPlaceholder = value.Bool;
+            }
+            else
+            {
+                // Temporary:
+                // no-op if cursor placeholder isn't supported. This means the request isn't coming from VS.
+                // Can remove this once VSCode starts using this endpoint for auto closing <text> tags.
+                return Task.FromResult(EmptyArray);
+            }
+
+            var formattingContext = CreateFormattingContext(uri, codeDocument, new Range(position, position), options);
+            if (IsAtTextTag(formattingContext, position))
+            {
+                // This is a text tag.
+                var cursorPlaceholder = addCursorPlaceholder ? LanguageServerConstants.CursorPlaceholderString : string.Empty;
```

Comment:
Should we just assume that we're always doing the placeholder? I ask because this auto-complete isn't super valuable without it

Reply:
This is the only way to differentiate VS and VSCode for now. We don't want to do placeholder for VSCode.